Case Tile Grouping UI

Jtang-1 opened the pull request:
## Product Description
This PR adds the UI for case tile grouping.
![Screen Shot 2023-06-21 at 7 57 47 PM](https://github.com/dimagi/commcare-hq/assets/88759246/1012bf48-a95a-445d-a342-b29a3d5168f5)

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Spec: https://docs.google.com/document/d/1me7aYlJTijDuUP7jC2-yJFwG86h8QUoOBSyWUOuU4uY/edit#
Jira: https://dimagi-dev.atlassian.net/browse/USH-3289
Related to: https://github.com/dimagi/commcare-hq/pull/33106

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
Case Tiles feature flags: https://confluence.dimagi.com/display/saas/Allow+Configuration+of+Case+List+Tiles

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Local testing. The change is limited to case tiles UI. None of the existing view functions are affected since new views are added for Grouped Case Tiles. 

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
No automated test
### QA Plan
<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Will perform a brief QA pre-deploy and do a more extensive QA after.
https://dimagi-dev.atlassian.net/browse/QA-5291
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
feat: get items from formplayer response

Jtang-1 committed:
feat: returns new CaseTileGroupedList view for menu

Jtang-1 committed:
feat: set up childView for CaseTileListView

Jtang-1 committed:
feat: maps child models to groupKeys and remove models to have only one childView per group

Jtang-1 committed:
feat: pass options to childView

Jtang-1 committed:
feat: gets header and row data and passes to template

Jtang-1 committed:
feat: display grouped case tiles on webapps

Jtang-1 committed:
feat: adds vertical spacing between group rows

Jtang-1 committed:
feat: adjust position of rows offshift from header

Jtang-1 committed:
feat: create white space between groups

Jtang-1 committed:
lint

Jtang-1 committed:
Merge branch 'master' into jt/case_tile_grouping_v1

Jtang-1 committed:
Merge branch 'sk/case-tile-group-v0' into jt/case_tile_grouping_v2

snopoke commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -306,6 +306,34 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
+        templateContext: function () {
+            const dict = CaseTileGroupedView.__super__.templateContext.apply(this, arguments);
+            dict['groupHeaderRows'] = this.options.groupHeaderRows;
+            dict['headerData'] = this.options.model.get('data').slice(0, this.options.groupHeaderRows);
```
This isn't quite right, this takes the first N `fields` however we want the first N `rows` (each row may have more than 1 field). For example, using the "Title row, second row with two cells, third row, and map" template with 2 header rows would require selecting the first 3 fields.

I'm not sure that we can guarantee that the fields will appear in their 'row' order. 

What I think we need to do is to use the grid configuration for each tile field to determine if it is part of the 'header' or not.

snopoke replied:
In the interest of time I took a stab at this: 83d92849fac

This takes all fields where `gridY < headerRows` i.e. all fields that are (at least partially) in the header rows.

I'm ignoring `gridHeight` (which could make a field span beyond the header) for now under the assumption that we won't configure things that way.

Jtang-1 replied:
Ah dang, thank you for the help implementing that as well. The first "N" row behavior is what i understood coming out our discussion previously but it somehow shifted while I was implementing.

snopoke commented on corehq/apps/cloudcare/templates/formplayer/case_list.html:
```diff
@@ -128,6 +128,53 @@
+    </div>
+  <% } %>
+
+  <div class = "group-data">
```
nit:
```suggestion
  <div class="group-data">
```

Jtang-1 replied:
[357737e](https://github.com/dimagi/commcare-hq/pull/33157/commits/357737eea68bc1b6753b45c69c8fab7df631f49f)

snopoke commented on corehq/apps/cloudcare/templates/formplayer/case_list.html:
```diff
@@ -128,6 +128,53 @@
+            <div class="webapp-markdown-output"><%= renderMarkdown(datum) %></div>
+          <% } %>
+        </div>
+      <% }); %>
```
Aside: It would be great if we didn't have to repeat all of this template. I was hoping you could do nested templates but it seems that would require a lot more work.

orangejenny replied:
Agreed. I don't think it needs to be part of this PR but do think it's worth a little investigation. Possibly a similar approach to `buildCaseTileStyles`, where the view's `initialize` function calls template functions, passing them only the data they need, and then injects the returned HTML strings into containers in the "parent" template?

Jtang-1 replied:
I think this portion can be pulled out as a template since `case-tile-view-item-template` and `case-tile-grouped-view-item-template` both header and rows all essentially use the same index,datum loop pattern. 
Then any specific html class and be injected via the view?
https://github.com/dimagi/commcare-hq/blob/1b35ebf340d2b451e5dcfb5b4b4e5dc3420ee0d1/corehq/apps/cloudcare/templates/formplayer/case_list.html#L114-L128

snopoke commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/collections.js:
```diff
@@ -54,6 +54,8 @@ hqDefine("cloudcare/js/formplayer/menus/collections", function () {
             'multiSelect',
             'maxSelectValue',
             'hasDetails',
+            'entities',
```
I don't think it's necessary to add 'entities' here.

Jtang-1 replied:
[357737e](https://github.com/dimagi/commcare-hq/pull/33157/commits/357737eea68bc1b6753b45c69c8fab7df631f49f)

snopoke reviewed (commented):
Few comments but looking good.

snopoke committed:
show all fields in header rows

Jtang-1 committed:
bug: row data is offset based on last header field index so that row data is in correct cell position

Jtang-1 committed:
lint and remove unneeded property from response

Jtang-1 committed:
bug: don't display group row if the row is empty

Jtang-1 committed:
bug: checks that tile.tile is not null (like when a property to configured as only sort) before accessing gridY property

Jtang-1 committed:
bug: filter hidden sort fields from populating and causing visually blank row

orangejenny commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -306,6 +306,41 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
+            return dict;
+        },
+
+        getIndexedRowDataList: function () {
```
How feasible would it be to write a test for this?

Jtang-1 replied:
I spent a chunk of time trying to do this, but having a hard time with the set up.

I don't know if there's a way to access `getIndexedRowDataList` directly from `CaseTileGroupedView` but i'm trying to get it through first initializing `CaseTileGroupedListView` which requires initializing all the inherited views and that's giving me trouble.

orangejenny replied:
fair enough

Jtang-1 committed:
refactor: use _groupby

Jtang-1 committed:
bug: creates indexed mapping instead of depending on offsetting index

Jtang-1 committed:
bug: maps index instead of relying on order
not all tiles are displayed so indices defined inheaderRowIndices are not necessarily
a consecutive sequence

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -755,6 +790,42 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
+        initialize: function () {
+            CaseTileGroupedListView.__super__.initialize.apply(this, arguments);
+
+            this.groupedModels = {};
```
Yes! [aa6528b](https://github.com/dimagi/commcare-hq/pull/33157/commits/aa6528b54170b309715fd88cf5e70327358654ec)

Jtang-1 committed:
refactor: move styling

Jtang-1 committed:
feat: align header and rows left edge

Jtang-1 commented on corehq/apps/hqwebapp/static/cloudcare/less/formplayer-webapp/menu.less:
```diff
@@ -38,6 +38,11 @@
   }
+
+  .case-tile-group{
+    margin: 10px 10px 10px;
```
[554b5e6](https://github.com/dimagi/commcare-hq/pull/33157/commits/554b5e6a461dbc0c74bfa16fa77aa19d0bf68435)

Jtang-1 commented on corehq/apps/cloudcare/templates/formplayer/case_list.html:
```diff
@@ -147,6 +192,19 @@
   justify-items: left;
   padding: 10px 20px 0 20px;
   }
+  .group-data{
```
good call [554b5e6](https://github.com/dimagi/commcare-hq/pull/33157/commits/554b5e6a461dbc0c74bfa16fa77aa19d0bf68435)

Jtang-1 commented:
Closing and reopening to run tests

Jtang-1 committed:
Merge branch 'master' into jt/case_tile_grouping_v2

orangejenny commented:
@Jtang-1 thank you for the followup!
